### DISCUSSION TITLE
Finding pattern with high periodicity and not just the least profile score

### DISCUSSION THREAD

**Discussion by kohinoor23** [TDAmeritrade/stumpy#906] [Help: Coding & Implementations]
stumpy provides a matrix profile based on nearest neighbour distance. I am interested to find out patterns that are well-defined and occur frequently. How can I ensure the occur frequently part using stumpy?

**Comment by JaKasb** [comment]
? Maybe search for Time Series Chains, and sort the chains by length ?
Your problem depends on the definition of "occur" and "well defined".
Maybe experiment with a threshold on the nearest-neighbor-distance to classify the sliding-windows into positive (occured) and negative categories.
It could be helpful to normalize the distances with CID , Complexity Invariant Distance Measure.

**Comment by seanlaw** [comment]
> stumpy provides a matrix profile based on nearest neighbour distance. I am interested to find out patterns that are well-defined and occur frequently. How can I ensure the occur frequently part using stumpy?

In terms of "well-defined", I'm assuming that you mean that the one nearest neighbor is "close enough" (and all other neighbors are either as close or farther away). So, you'll need to choose a one nearest neighbor distance that suits your criteria and then extract all of those corresponding motif subsequences (i.e., there may be, say, 20 subsequences with a one nearest neighbor that is within a distance of 5.5 and so you extract those 20 subsequences).

> How can I ensure the occur frequently part using stumpy?

Now, for "frequency", you can take the 20 subsequences from above and search for all possible matches using [stumpy.match](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.match). With this function you can also specify some matching conditions and then be able to count how many matches meet that your desired matching criteria. It is entirely possible that your criteria is too stringent and your data does not have any matches but that is not a matrix profile problem but the nature of your data.

**Comment by NimaSarajpoor** [comment]
@atul2602  [another way to think about your problem]

Matrix Profile gives you 1NN distance. This is the same as finding the minimum value per row / col in a distance matrix, which contains pairwise distances.  So, let's say you have the distance matrix that gives you the distance between any two subsequence of your time series. For EACH subsequence, you can now find top-10 nearest neighbors. 

So, here is the question:
How do you say if "a subsequence and its top-10 neighbors" is better than "another subsequence and its top-10 neighbors"?
